### PR TITLE
ceph-bluestore-tool: Fixes for multilple bdev label

### PR DESCRIPTION
**Pull request by aclamk**
This PR consists of:

ZAP:
https://github.com/ceph/ceph/pull/59632
show-label:
https://github.com/ceph/ceph/pull/59714
really-really-mean-it:
https://github.com/ceph/ceph/pull/59627
+
-zap-size param is removed
+
documentation added

Fixes:
https://tracker.ceph.com/issues/68048

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: doc/man/8/ceph-bluestore-tool.rst [hunk 1]
```diff
@@ -106,7 +107,10 @@ Commands
 
 :command:`show-label` --dev *device* [...]
 
-   Show device label(s).	   
+   Show device label(s). Label is a special 4K area that plays the role of superblock,
+   provides metadata that exposes version, creation time, size, etc.
+   There are multiple copies of the block at offsets: 0, 1G, 10G, 100G and 1000G.
+   As an exception, label can be printed even if an OSD is running.
```
Comment (anthonyeleven) [inline doc/man/8/ceph-bluestore-tool.rst:113]: The label is a 4KB structure at the beginning of the device that serves as a superblock, holding metadata including version, creation time, size, etc.  There are copies at offsets 1GB, 10GB, 100GB, and 1000GB.  The label may be printed while an OSD is running.

Comment (anthonyeleven) [reply doc/man/8/ceph-bluestore-tool.rst:113]: Isn't this redundant wrt existing text around lines 203 / 211?

Comment (guits) [reply doc/man/8/ceph-bluestore-tool.rst:113]: it feels to me this change should actually be moved to section `Device labels`

Comment (aclamk) [reply doc/man/8/ceph-bluestore-tool.rst:113]: Right.
I will move detailed explanation around "Device labels" section.

Comment (aclamk) [reply doc/man/8/ceph-bluestore-tool.rst]: @anthonyeleven Please recheck. My PRs always benefit from your language input.

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -8992,27 +8992,20 @@ void BlueStore::trim_free_space(const string& type, std::ostream& outss)
...
-int BlueStore::zap_device(CephContext* cct, const string& dev, uint64_t gap_size)
+int BlueStore::zap_device(CephContext* cct, const string& dev)
 {
-  string path; // dummy var for dout
-  dout(5)<<  __func__ << " " << dev << dendl;
+  string path = dev; // dummy var for dout
+  uint64_t brush_size;
+  dout(5) << __func__ << " " << dev << dendl;
   auto * _bdev =
     BlockDevice::create(cct, dev, nullptr, nullptr, nullptr, nullptr);
```
Comment (YiteGu) [inline src/os/bluestore/BlueStore.cc]: suggest using smart pointers here:
`unique_ptr<BlockDevice> _bdev(BlockDevice::create(cct, dev, nullptr, nullptr, nullptr, nullptr));`

**Review by guits — APPROVED** [review]
tested with a cephadm deployed cluster:

`show-label` on running OSD:
```
[root@bluestore-tests-1 /]# ceph osd tree
ID  CLASS  WEIGHT   TYPE NAME                   STATUS  REWEIGHT  PRI-AFF
-1         0.19530  root default
-3         0.19530      host bluestore-tests-1
 0    hdd  0.19530          osd.0                   up   1.00000  1.00000
[root@bluestore-tests-1 /]# ceph-bluestore-tool show-label --dev /dev/ceph-f895bb03-9590-46e7-be60-f2aa39965576/osd-block-e06e7091-ad93-436b-8cff-9bc45bebe672
{
    "/dev/ceph-f895bb03-9590-46e7-be60-f2aa39965576/osd-block-e06e7091-ad93-436b-8cff-9bc45bebe672": {
        "osd_uuid": "e06e7091-ad93-436b-8cff-9bc45bebe672",
        "size": 214744170496,
        "btime": "2024-09-13T08:29:22.110570+0000",
        "description": "main",
        "bfm_blocks": "52427776",
        "bfm_blocks_per_key": "128",
        "bfm_bytes_per_block": "4096",
        "bfm_size": "214744170496",
        "bluefs": "1",
        "ceph_fsid": "efc6b532-71a9-11ef-bf2c-525400c3b0c0",
        "ceph_version_when_created": "ceph version 19.3.0-4916-g3e18128d (3e18128dc8e02730bc1b3e7530b4fbb543f81eef) squid (dev)",
        "created_at": "2024-09-13T08:29:25.624292Z",
        "elastic_shared_blobs": "1",
        "epoch": "18",
        "kv_backend": "rocksdb",
        "magic": "ceph osd volume v026",
        "multi": "yes",
        "osd_key": "AQDg9+NmJWGcMhAA99mv/fZte7DwR/L+9X00jA==",
        "osdspec_affinity": "None",
        "ready": "ready",
        "require_osd_release": "19",
        "type": "bluestore",
        "whoami": "0"
    }
}
[root@bluestore-tests-1 /]#
```

`ceph-bluestore-tool zap-device`:
```
[root@bluestore-tests-1 /]# ceph orch daemon stop osd.0
Scheduled to stop osd.0 on host 'bluestore-tests-1'

[root@bluestore-tests-1 /]# ceph osd tree
ID  CLASS  WEIGHT   TYPE NAME                   STATUS  REWEIGHT  PRI-AFF
-1         0.19530  root default
-3         0.19530      host bluestore-tests-1
 0    hdd  0.19530          osd.0                 down   1.00000  1.00000

[root@bluestore-tests-1 /]# hexdump -C -n22 -s1073741824 /dev/ceph-f895bb03-9590-46e7-be60-f2aa39965576/osd-block-e06e7091-ad93-436b-8cff-9bc45bebe672
40000000  62 6c 75 65 73 74 6f 72  65 20 62 6c 6f 63 6b 20  |bluestore block |
40000010  64 65 76 69 63 65                                 |device|
40000016

[root@bluestore-tests-1 /]# ceph-bluestore-tool zap-device --dev /dev/ceph-f895bb03-9590-46e7-be60-f2aa39965576/osd-block-e06e7091-ad93-436b-8cff-9bc45bebe672 --yes-i-really-really-mean-it

[root@bluestore-tests-1 /]# hexdump -C -n22 -s1073741824 /dev/ceph-f895bb03-9590-46e7-be60-f2aa39965576/osd-block-e06e7091-ad93-436b-8cff-9bc45bebe672
40000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
40000010  00 00 00 00 00 00                                 |......|
40000016
[root@bluestore-tests-1 /]#
```

**Comment by aclamk** [comment]
https://pulpito.ceph.com/akupczyk-2024-09-13_09:37:56-rados-wip-aclamk-cbt-combined-distro-default-smithi/
https://pulpito.ceph.com/akupczyk-2024-09-13_09:43:51-rados-wip-aclamk-cbt-combined-distro-default-smithi/
https://pulpito.ceph.com/akupczyk-2024-09-13_10:00:18-upgrade-wip-aclamk-cbt-combined-distro-default-smithi/

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc]: Good one. Done.

File: doc/man/8/ceph-bluestore-tool.rst [hunk 3]
```diff
@@ -30,6 +30,7 @@ Synopsis
 | **ceph-bluestore-tool** reshard --path *osd path* --sharding *new sharding* [ --sharding-ctrl *control string* ]
 | **ceph-bluestore-tool** show-sharding --path *osd path*
 | **ceph-bluestore-tool** trim --path *osd path*
+| **ceph-bluestore-tool** zap --path *osd path*
```
Comment (ifed01) [inline doc/man/8/ceph-bluestore-tool.rst:33]: That's not <osd path> but <dev path>.

File: doc/man/8/ceph-bluestore-tool.rst [hunk 4]
```diff
@@ -139,6 +141,10 @@ Commands
    and allows the drive to perform more efficient internal housekeeping.
    If BlueStore runs with discard enabled, this option may not be useful.
 
+:command: `zap` --path *osd path*
```
Comment (ifed01) [inline doc/man/8/ceph-bluestore-tool.rst:144]: same here

Comment (guits) [reply doc/man/8/ceph-bluestore-tool.rst:144]: by the way, shouldn't it be `zap-device` ?
```suggestion
:command: `zap-device` --path *osd path*
```

Comment (aclamk) [reply doc/man/8/ceph-bluestore-tool.rst]: Fixed.

File: doc/man/8/ceph-bluestore-tool.rst (hunk 3 above)
Comment (aclamk) [reply doc/man/8/ceph-bluestore-tool.rst]: Fixed.

**Review by ifed01 — APPROVED** [review]

**Comment by YiteGu** [comment]
jenkins test make check

**Comment by guits** [comment]
jenkins test api

**Comment by guits** [comment]
https://pulpito.ceph.com/gabrioux-2024-09-16_07:44:22-orch:cephadm-wip-guits-main-2024-09-13-1248-distro-default-smithi/
https://pulpito.ceph.com/gabrioux-2024-09-16_13:13:56-orch:cephadm-wip-guits-main-2024-09-13-1248-distro-default-smithi/

**Comment by aclamk** [comment]
https://pulpito.ceph.com/akupczyk-2024-09-18_21:20:02-upgrade:reef-x-aclamk-testing-nauvoo-2024-09-18-1004-distro-default-smithi/
https://pulpito.ceph.com/akupczyk-2024-09-18_17:28:43-rados-aclamk-testing-nauvoo-2024-09-18-1004-distro-default-smithi/
https://pulpito.ceph.com/akupczyk-2024-09-18_17:27:22-rados-aclamk-testing-nauvoo-2024-09-18-1004-distro-default-smithi/

**Pull request merged**